File: datasets/Part 9 - Dimensionality Reduction/Section 45 - Kernel PCA/kernel_acp.R
# Kernel PCA

# =======================================================================================================
#
# Permite reducir la dimension cuando se trata con un problema de Clasificacion que NO es linealmente
# separable. Es una version de PCA adaptada con un kernel que transforma los datos, primero a una dimension
# superior donde si sean linealmente separables y a partir de ahi exptraer componentes principales para
# solucionar el problema como si fuera linelamente separable.
#
# Como primero se aplico un kernel y luego se hizo un PCA a un conjunto de datos que no era linealmente
# searable, ahora las observaciones se localizan en zonas del plano diferentes a las que se tenian en un
# principio.
# Los datos se separan de una mejor manera por una recta que sin aplicar el kernel.
#
# =======================================================================================================


################################################
###          IMPORTAR EL DATA SET            ###
################################################

setwd("~/Documentos/Udemy/machinelearning-az/datasets/Part 9 - Dimensionality Reduction/Section 45 - Kernel PCA")
dataset = read.csv('Social_Network_Ads.csv')
dataset = dataset[, 3:5]

#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

# install.packages("caTools")
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

################################################
#            Escalado de variables             #
################################################

training_set[,1:2] = scale(training_set[,1:2])
testing_set[,1:2] = scale(testing_set[,1:2])

#######################################################
#   Reducir la dimensión del dataset con Kernel PCA   #
#######################################################

#install.packages("kernlab")
library(kernlab)
# Se crea el kernel que eumentara la dimension de los datos para hacerlos linealmente separables.
# La formula se puede escribir como "~." para indicar que todas las variables son independientes,
# por lo que del trining set se quita la columna de la variable dependiente.
# Se puede aplicar un kernel gaussiano (rbf), polinoial, sigmoide, etc. dependiendo de la naturaleza de los datos.
# con "features", seleccionamos el no. de componentes principales a tomar en cuenta.
kpca = kpca(~., data = training_set[, -3], 
            kernel = "rbfdot", features = 2)
# Aplicamos la transformacion generada en el paso anteroir al dataset de Entrenamiento y de Testing
# Y cambiamos el orden de las columnas.
# Los datasets de Training y de Testing, ahora tendran tantas columnas como componentes principales se hayan elegido.
training_set_pca = as.data.frame(predict(kpca, training_set)) # se necesita hacer un casting a DataFrame
# Se le añade al training set la columna de la variable dependiente
training_set_pca$Purchased = training_set$Purchased
testing_set_pca = as.data.frame(predict(kpca, testing_set)) # se necesita hacer un casting a DataFrame
# Se le añade al testing set la columna de la variable dependiente
testing_set_pca$Purchased = testing_set$Purchased

# training_set - Tiene al dataset con las variables escaladas.
# training_set_pca - Tiene al dataset con las variables escaladas y con la transformacion.

#########################################################
#    Ajustar el modelo con el dataset de Entrenamiento  #
#########################################################

# En esta parte se agrega el modelo de clasificacion que se decida.

classifier = glm(formula = Purchased ~ .,
                 data = training_set_pca, 
                 family = binomial)

################################################
#                PREDICCION                    #
################################################

# Predicción de los resultados con el conjunto de testing
prob_pred = predict(classifier, type = "response",
                    newdata = testing_set_pca[,-3])

y_pred = ifelse(prob_pred> 0.5, 1, 0)

################################################
#        EVALUACION DEL RENDIMIENTO            #
################################################

# Crear la matriz de confusión
cm = table(testing_set_pca[, 3], y_pred)

################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

# Visualización del conjunto de entrenamiento
#install.packages("ElemStatLearn")
library(ElemStatLearn)
set = training_set_pca
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('V1', 'V2')
prob_set = predict(classifier, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[, -3],
     main = 'Clasificación (Conjunto de Entrenamiento)',
     xlab = 'CP1', ylab = 'CP2',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))


# Visualización del conjunto de testing
set = testing_set_pca
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('V1', 'V2')
prob_set = predict(classifier, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[, -3],
     main = 'Clasificación (Conjunto de Testing)',
     xlab = 'CP1', ylab = 'CP2',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

File: datasets/Part 9 - Dimensionality Reduction/Section 45 - Kernel PCA/kernel_acp.py
#!/home/kevinml/anaconda3/bin/python3.7
# -*- coding: utf-8 -*-
"""
Created on Thu May  2 10:50:27 2019

@author: juangabriel and Kevin Meza
"""

# Kernel PCA

# =======================================================================================================
#
# Permite reducir la dimension cuando se trata con un problema de Clasificacion que NO es linealmente
# separable. Es una version de PCA adaptada con un kernel que transforma los datos, primero a una dimension
# superior donde si sean linealmente separables y a partir de ahi exptraer componentes principales para
# solucionar el problema como si fuera linelamente separable.
#
# Como primero se aplico un kernel y luego se hizo un PCA a un conjunto de datos que no era linealmente
# searable, ahora las observaciones se localizan en zonas del plano diferentes a las que se tenian en un
# principio.
# Los datos se separan de una mejor manera por una recta que sin aplicar el kernel.
#
# =======================================================================================================

# Cómo importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

################################################
###          IMPORTAR EL DATA SET            ###
################################################

dataset = pd.read_csv('Social_Network_Ads.csv')

X = dataset.iloc[:, [2, 3]].values
y = dataset.iloc[:, 4].values


#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=0)


################################################
#            Escalado de variables             #
################################################

from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

#######################################################
#   Reducir la dimensión del dataset con Kernel PCA   #
#######################################################

from sklearn.decomposition import KernelPCA
# Con el parametro "n_components", seleccionamos el no. de componentes principales,
# si el valor de este parametro es "None", se calcula un analisis con todas las componentes principales y posteriormente se puede
# visualizar que porcentaje de la varianza explica cada componente.
# Se puede aplicar un kernel gaussiano (rbf), polinoial, sigmoide, etc. dependiendo de la naturaleza de los datos.
kpca = KernelPCA(n_components=2, kernel="rbf")
X_train = kpca.fit_transform(X_train)
X_test = kpca.transform(X_test)
# En esta parte se crea una variabe que contendra cada uno de los componentes principales y la varianza que explica cada uno de estos.
explained_variance = pca.explained_variance_ratio_


#########################################################
#    Ajustar el modelo con el dataset de Entrenamiento  #
#########################################################

# En esta parte se agrega el modelo de clasificacion que se decida.

from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train, y_train)

################################################
#                PREDICCION                    #
################################################

# Predicción de los resultados con el Conjunto de Testing
y_pred = classifier.predict(X_test)

################################################
#        EVALUACION DEL RENDIMIENTO            #
################################################

# Elaborar una matriz de confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

# Representación gráfica de los resultados del algoritmo en el Conjunto de Entrenamiento
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start=X_set[:, 0].min() - 1, stop=X_set[:, 0].max() + 1, step=0.01),
                     np.arange(start=X_set[:, 1].min() - 1, stop=X_set[:, 1].max() + 1, step=0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha=0.75, cmap=ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c=ListedColormap(('red', 'green'))(i), label=j)
plt.title('Clasificador (Conjunto de Entrenamiento)')
plt.xlabel('CP1')
plt.ylabel('CP2')
plt.legend()
plt.show()


# Representación gráfica de los resultados del algoritmo en el Conjunto de Testing
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start=X_set[:, 0].min() - 1, stop=X_set[:, 0].max() + 1, step=0.01),
                     np.arange(start=X_set[:, 1].min() - 1, stop=X_set[:, 1].max() + 1, step=0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha=0.75, cmap=ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c=ListedColormap(('red', 'green'))(i), label=j)
plt.title('Clasificador (Conjunto de Test)')
plt.xlabel('CP1')
plt.ylabel('CP2')
plt.legend()
plt.show()
